From 94f0bfae10cef40291b1096a49206b15bc8eb45d Mon Sep 17 00:00:00 2001
From: Andrew Yourtchenko <ayourtch@gmail.com>
Date: Wed, 19 Jun 2019 12:09:51 +0000
Subject: [PATCH] interface: fix the inforrect sizes/offsets in the tso segmentation

The copying of the first segment copied too small amount of data.
The copying of the second and subsequent segments used the wrong
data offset - for the case of GSO enabled it starts from 12 bytes
earlier.

Change-Id: I3adc532c175babc1ca1e121c7e12e6cafbdb9974
Type: fix
Ticket: VPP-1700
Signed-off-by: Andrew Yourtchenko <ayourtch@gmail.com>
---

diff --git a/src/vnet/interface_output.c b/src/vnet/interface_output.c
index 2e4f360..4b2ca2c 100644
--- a/src/vnet/interface_output.c
+++ b/src/vnet/interface_output.c
@@ -352,7 +352,7 @@
 
   vlib_buffer_t *b0 = vlib_get_buffer (vm, ptd->split_buffers[0]);
   tso_init_buf_from_template_base (b0, sb0, default_bflags,
-				   l4_hdr_sz + first_data_size);
+				   l234_sz + first_data_size);
 
   u32 total_src_left = n_bytes_b0 - l234_sz - first_data_size;
   if (total_src_left)
@@ -410,7 +410,7 @@
 		  csbi0 = next_bi;
 		  csb0 = vlib_get_buffer (vm, csbi0);
 		  src_left = csb0->current_length;
-		  src_ptr = csb0->data;
+		  src_ptr = vlib_buffer_get_current (csb0);
 		}
 	      else
 		{
